### PR TITLE
VisualizationView improvements and bug fixes.

### PR DESCRIPTION
**Pull request by damonkohler**
- Fixes race condition bugs that can lead to NPEs.
- Changes Layer interface to hand in VisualizationView everywhere. This cuts down on the number of parameters to methods like onStart().
- Adds new methods to Layer interface to make it easier to preload assets like fonts.
- Adds TextShape and PoiShape.
- Fixes bug in OccupancyGridLayer. The texture was not unbound after rendering.
- Changes XYOrthographicCamera to separate the ROS to screen transform and the camera transform. This was necessary to enable rendering text aligned with the screen.

### CODE REVIEW

**Comment by damonkohler** [comment]
PTAL @SirVer

File: android_15/src/org/ros/android/view/visualization/VisualizationView.java [hunk 1]
```diff
@@ -46,34 +49,50 @@
...
-  private void init() {
+  /**
+   * Must be called in {@link Activity#onCreate(Bundle)}.
+   * 
+   * @param layers
+   */
+  public void onCreate(List<Layer> layers) {
+    Preconditions.checkNotNull(layers);
+    this.layers = layers;
+    setDebugFlags(DEBUG_CHECK_GL_ERROR);
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/VisualizationView.java:74]: should this not only happen on DEBUG == true?

File: android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java [hunk 2]
```diff
@@ -38,23 +38,36 @@
    * meter (the display density) then 1 cm in the world will be displayed as 1
    * cm on the display.
    */
-  private static final double DEFAULT_ZOOM = 100.0;
+  private static final double PIXELS_PER_METER = 100.0;
+
+  /**
+   * Transforms from our ROS frame (the data frame) to the screen frame (our
+   * target frame). Rotate coordinate system to match ROS standard (x is
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java]: Rotates coordinate ...

File: android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java [hunk 3]
```diff
@@ -95,61 +108,83 @@ public boolean applyFrameTransform(GL10 gl, GraphName frame) {
...
-   * origin is the top left corner of the view
+   * @return the provided pixel coordinates (where the origin is the top left
+   *         corner of the view) in {@link #frame}
    */
-  public Vector3 toMetricCoordinates(int x, int y) {
-    double centeredX = x - viewport.getWidth() / 2.0d;
-    double centeredY = viewport.getHeight() / 2.0d - y;
-    return transform.invert().apply(new Vector3(centeredX, centeredY, 0));
+  public Vector3 toCameraFrame(int pixelX, int pixelY) {
+    double centeredX = pixelX - viewport.getWidth() / 2.0d;
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java]: can both be final.

File: android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java [hunk 4]
```diff
@@ -188,15 +224,16 @@ public void setFrame(String frame) {
...
-   * @param frame the new camera frame
+   * 
+   * @param frame
+   *          the new camera frame
    */
   public void jumpToFrame(GraphName frame) {
     synchronized (mutex) {
       this.frame = frame;
-      double zoom = getZoom();
+      double scale = cameraToRosTransform.getScale();
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java]: final?

File: android_15/src/org/ros/android/view/visualization/XYOrthographicRenderer.java [hunk 5]
```diff
@@ -35,66 +32,55 @@
...
   public void onSurfaceChanged(GL10 gl, int width, int height) {
     Viewport viewport = new Viewport(width, height);
     viewport.apply(gl);
-    camera.setViewport(viewport);
+    view.getCamera().setViewport(viewport);
     gl.glMatrixMode(GL10.GL_MODELVIEW);
     gl.glEnable(GL10.GL_BLEND);
     gl.glBlendFunc(GL10.GL_SRC_ALPHA, GL10.GL_ONE_MINUS_SRC_ALPHA);
     gl.glDisable(GL10.GL_DEPTH_TEST);
+    // TODO(damonkohler): Pull out a constant for this?
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicRenderer.java]: yes please.

File: android_15/src/org/ros/android/view/visualization/VisualizationView.java (hunk 1 above)
Comment (damonkohler) [inline android_15/src/org/ros/android/view/visualization/VisualizationView.java:74]: No, I think it's best to enable it permanently. There's probably some performance impact, but nothing that isn't swamped by other stuff.

File: android_15/src/org/ros/android/view/visualization/shape/PixelSpacePoseShape.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Represents a pose.
+ * <p>
+ * This shape is defined in pixel space and will not be affected by the zoom
+ * level of the camera.
+ * 
+ * @author damonkohler@google.com (Damon Kohler)
+ */
+public class PixelSpacePoseShape extends MetricSpacePoseShape {
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/shape/PixelSpacePoseShape.java:31]: I somewhat dislike this name, but I cannot come up with a better one. NonzoomingPoseShape? FixedSizePoseShape?

File: android_15/src/org/ros/android/view/visualization/shape/TextShape.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
+package org.ros.android.view.visualization.shape;
+
+import org.ros.android.view.visualization.VisualizationView;
+import uk.co.blogspot.fractiousg.texample.GLText;
+
+import javax.microedition.khronos.opengles.GL10;
+
+public class TextShape extends BaseShape {
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/shape/TextShape.java:8]: add javadoc.

File: android_15/src/org/ros/android/view/visualization/shape/TextShapeFactory.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package org.ros.android.view.visualization.shape;
+
+import org.ros.android.view.visualization.VisualizationView;
+import uk.co.blogspot.fractiousg.texample.GLText;
+
+import javax.microedition.khronos.opengles.GL10;
+
+public class TextShapeFactory {
```
Comment (SirVer) [inline android_15/src/org/ros/android/view/visualization/shape/TextShapeFactory.java:8]: Add javadoc.

File: android_15/src/uk/co/blogspot/fractiousg/texample/GLText.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
+// This is a OpenGL ES 1.0 dynamic font rendering system. It loads actual font
```
Comment (SirVer) [inline android_15/src/uk/co/blogspot/fractiousg/texample/GLText.java:1]: Author, license and link?

**Comment by SirVer** [comment]
bunch of small comments.

File: android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java (hunk 2 above)
Comment (damonkohler) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java]: Done.

File: android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java (hunk 3 above)
Comment (damonkohler) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java]: Done.

File: android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java (hunk 4 above)
Comment (damonkohler) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicCamera.java]: Done.

File: android_15/src/org/ros/android/view/visualization/XYOrthographicRenderer.java (hunk 5 above)
Comment (damonkohler) [inline android_15/src/org/ros/android/view/visualization/XYOrthographicRenderer.java]: Done.

**Pull request merged**